sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -648,30 +643,55 @@ static status_t kmac_process_msg_blocks(kmac_operation_t operation,
       abs_mmio_read32(kKmacBaseAddr + KMAC_CFG_SHADOWED_REG_OFFSET);
   uint32_t keccak_str =
       bitfield_field32_read(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD);
-  size_t keccak_rate;
-  HARDENED_TRY(kmac_get_keccak_rate_bytes(keccak_str, &keccak_rate));
-
-  // Finally, we can read the two shares of digest and XOR them
-  // Here the counter i denotes the number of bytes read from Keccak state
-  for (i = 0; i < digest->len; i++) {
-    // Do we require additional Keccak rounds?
-    if ((i % keccak_rate) == 0 && i > 0) {
-      // if we consumed all Keccak state and aren't done yet, run one more
-      // Keccak round
+  size_t keccak_rate_words;
+  HARDENED_TRY(kmac_get_keccak_rate_words(keccak_str, &keccak_rate_words));
+
+  // Finally, we can read the two shares of digest and XOR them.
+  size_t idx = 0;
+
+  size_t digest_len_words = digest_len / sizeof(uint32_t);
+  if (digest_len % sizeof(uint32_t) != 0) {
+    digest_len_words++;
+  }
+  uint32_t digest_words[digest_len_words];
+  while (launder32(idx) < digest_len_words) {
+    // Since we always read in increments of the Keccak rate, the index at
+    // start should always be a multiple of the rate.
+    HARDENED_CHECK_EQ(idx % keccak_rate_words, 0);
+
+    // Poll the status register until in the 'squeeze' state.
+    HARDENED_TRY(wait_status_bit(KMAC_STATUS_SHA3_SQUEEZE_BIT, 1));
+
+    // Read words from the state registers (either `digest_len_words` or the
+    // maximum number of words available).
+    size_t offset = 0;
+    for (; launder32(idx) < digest_len_words && offset < keccak_rate_words;
+         offset++) {
+      uint32_t share0 =
+          abs_mmio_read32(kKmacStateShare0Addr + offset * sizeof(uint32_t));
+      uint32_t share1 =
+          abs_mmio_read32(kKmacStateShare1Addr + offset * sizeof(uint32_t));
+      digest_words[idx] = share0 ^ share1;
+      ++idx;
+    }
```

Comment:
I agree, but this is the digest rather than the key; since it's ultimately returned and handled in unmasked form, I assumed that handling the shares close together here would be okay. What do you think?

Reply:
You're right, I confused this with the key. I don't think it's a big issue with the digest.